spec/integration/client/exit_code_spec.rb

```diff
@@ -0,0 +1,275 @@
+local_mode true
+cookbook_path "#{path_to('cookbooks')}"
+EOM
+    end
+
+    def setup_client_rb_with_audit_mode
+      file "config/client.rb", <<EOM
+local_mode true
+cookbook_path "#{path_to('cookbooks')}"
+audit_mode :audit_only
+EOM
+    end
+
+    def run_chef_client_and_expect_exit_code(exit_code)
+      shell_out!(
+        "#{chef_client} -c \"#{path_to('config/client.rb')}\" -o 'x::default'",
+        :cwd => chef_dir,
+        :returns => [exit_code])
+    end
+
+    context "has a cookbook" do
+
+      context "with a library" do
+
+        context "which cannot be loaded" do
+          before do
+            file "cookbooks/x/recipes/default.rb", ""
+            file "cookbooks/x/libraries/error.rb", "require 'does/not/exist'"
+          end
+
+          it "exits with GENERAL_FAILURE, 1" do
+            setup_client_rb
+            run_chef_client_and_expect_exit_code 1
+          end
+        end
+
+      end
+
+      context "with an audit recipe" do
+
+        context "which fails" do
+          before do
+            file "cookbooks/x/recipes/default.rb", <<-RECIPE
+control_group "control group without top level control" do
+  it "should fail" do
+    expect(2 - 2).to eq(1)
+  end
+end
+RECIPE
+          end
+
+          it "exits with GENERAL_FAILURE, 1" do
+            setup_client_rb_with_audit_mode
+            run_chef_client_and_expect_exit_code 1
+          end
+        end
+
+      end
+
+      context "with a recipe" do
+
+        context "which throws an error" do
+          before { file "cookbooks/x/recipes/default.rb", "raise 'BOOM'" }
+
+          it "exits with GENERAL_FAILURE, 1" do
+            setup_client_rb
+            run_chef_client_and_expect_exit_code 1
+          end
+        end
+
+        context "with a recipe which calls Chef::Application.fatal with a non-RFC exit code" do
+          before { file "cookbooks/x/recipes/default.rb", "Chef::Application.fatal!('BOOM', 123)" }
+
+          it "exits with the specified exit code" do
+            setup_client_rb
+            run_chef_client_and_expect_exit_code 123
+          end
+        end
+
+        context "with a recipe which calls Chef::Application.exit with a non-RFC exit code" do
+          before { file "cookbooks/x/recipes/default.rb", "Chef::Application.exit!('BOOM', 231)" }
+
+          it "exits with the specified exit code" do
+            setup_client_rb
+            run_chef_client_and_expect_exit_code 231
+          end
+        end
+
+        context "with a recipe that requests a reboot" do
+          before do
+            file "cookbooks/x/recipes/default.rb", <<EOM
+reboot "test" do
+  delay_mins 20
+  reason "functional test of reboot exit code"
+  action :reboot_now
+end
+EOM
+          end
+
+          after do
+            if Chef::Platform.windows?
+              shell_out!("shutdown /a")
+            else
+              shell_out!("shutdown -c")
+            end
+          end
+
+          xit "exits with SUCCESS, 0" do
```

Comment:
i think it's clearer to do `it "...", pending: true` than `xit`; easier to grep for too.